### PR TITLE
adding action 'set-vpc-flow`

### PR DESCRIPTION
**Pull request by joshuaroot**
- adding action `create-flow-logs` to `vpc.py`

### CODE REVIEW

**Comment by kapilt** [comment]
in general its always good to have actions cover on multiple use cases, ie. have as a set-vpc-flow vs a create.

File: c7n/resources/vpc.py [hunk 1]
```diff
@@ -1600,3 +1600,73 @@ class resource_type(object):
...
+            'DeliverLogsPermissionArn': {'type': 'string'},
+            'LogGroupName': {'type': 'string'},
+            'TrafficType': {'type': 'string',
+                            'enum': ['ACCEPT', 'REJECT', 'ALL']},
+            'required': ['DeliverLogsPermissionArn', 'LogGroupName',
+                         'ResourceIds']
+        }
+    }
+
+    def validate(self):
```
Comment (kapilt) [inline c7n/resources/vpc.py]: this validate method appears to be superflous given the jsonschema for the action.

File: c7n/resources/vpc.py [hunk 2]
```diff
@@ -1600,3 +1600,73 @@ class resource_type(object):
...
+        model = self.manager.get_model()
+        r_type = model.type
+        if r_type == 'vpc':
+            r_type = 'VPC'
+        elif r_type == 'subnet':
+            r_type = 'Subnet'
+        else:
+            r_type = 'NetworkInterface'
+        params['ResourceType'] = r_type
+        params['TrafficType'] = params['TrafficType'].upper()
```
Comment (kapilt) [inline c7n/resources/vpc.py]: flow log traffic type should default to ALL and not be required in schema.

File: c7n/resources/vpc.py [hunk 3]
```diff
@@ -1600,3 +1600,73 @@ class resource_type(object):
...
+        params['ResourceIds'] = [r[model.id] for r in resources]
+
+        client.create_flow_logs(**params)
```
Comment (kapilt) [inline c7n/resources/vpc.py]: The state absent case isn't being handled here.

**Comment by kapilt** [comment]
the previous review comments on this pr are still valid, the state: absent case needs implementation, and there doesn't need a separate recording/test case per applicable resource type.

File: c7n/resources/vpc.py [hunk 4]
```diff
@@ -1600,3 +1600,79 @@ class resource_type(object):
...
+        client.delete_flow_logs(FlowLogIds=[f['FlowLogId'] for f in flow_logs])
+
+    def process(self, resources):
+        client = local_session(self.manager.session_factory).client('ec2')
+        params = dict(self.data)
+        params.pop('type')
+
+        model = self.manager.get_model()
+        r_type = model.type
+        if r_type == 'vpc':
```
Comment (kapilt) [inline c7n/resources/vpc.py]: its minor but I think this type of lookup conditional is more nicely done as a map/dict lookup. ie. class var for {r_type: param_type}

File: c7n/resources/vpc.py [hunk 5]
```diff
@@ -1600,3 +1600,79 @@ class resource_type(object):
...
+        elif r_type == 'subnet':
+            r_type = 'Subnet'
+        else:
+            r_type = 'NetworkInterface'
+
+        params['ResourceType'] = r_type
+        params['TrafficType'] = self.data.get('TrafficType', 'ALL').upper()
+        params['ResourceIds'] = [r[model.id] for r in resources]
+
+        if not params['ResourceIds']:
```
Comment (kapilt) [inline c7n/resources/vpc.py]: this conditional can be elided/removed, policies execution won't invoke an action with an empty set of resources.

File: c7n/resources/vpc.py [hunk 6]
```diff
@@ -1600,3 +1600,79 @@ class resource_type(object):
...
+            self.log.info('No %s resources to set' % model.type)
+            return
+
+        if not params['LogGroupName']:
+            self.delete_flow_logs(client, params['ResourceIds'])
```
Comment (kapilt) [inline c7n/resources/vpc.py]: implicitly deleting resources is always a bit sketchy, I'd want a user to be pretty explicit here, ie state: absent|false which mirrors some of our other policy action constructions before we delete. forgetting a parameter, shouldn't lead to nuking things. also either a validate or

File: c7n/resources/vpc.py [hunk 7]
```diff
@@ -1600,3 +1600,79 @@ class resource_type(object):
...
+        if not self.data.get('DeliverLogsPermissionArn'):
+            if self.data.get('LogGroupName'):
+                raise ValueError('DeliverLogsPermissionArn required when '
+                                 'specifying LogGroupName')
+        return self
+
+    def delete_flow_logs(self, client, rids):
+        flow_logs = client.describe_flow_logs(
+            Filters=[{'Name': 'resource-id', 'Values': rids}])['FlowLogs']
+        client.delete_flow_logs(FlowLogIds=[f['FlowLogId'] for f in flow_logs])
```
Comment (kapilt) [inline c7n/resources/vpc.py]: we'd want to catch on the exception condition that the flow log isn't present and pass.

File: c7n/resources/vpc.py [hunk 8]
```diff
@@ -1600,3 +1600,79 @@ class resource_type(object):
...
+
+        if not params['ResourceIds']:
+            self.log.info('No %s resources to set' % model.type)
+            return
+
+        if not params['LogGroupName']:
+            self.delete_flow_logs(client, params['ResourceIds'])
+            return
+
+        client.create_flow_logs(**params)
```
Comment (kapilt) [inline c7n/resources/vpc.py]: if a flow log with different parameters exists, what happens when we make this api call?

Comment (joshuaroot) [reply c7n/resources/vpc.py]: so, if a flow-log with different parameters exist, a second destination is created...

**Pull request merged**